File: packages/interocitor-python/src/interocitor/recovery.py
"""Recovery-wrapper primitives compatible with ``@interocitor/core``.

Recovery phrases derive an opaque locator and an AES-GCM key-encryption key.
The wrapper contains only public KDF parameters and ciphertext; the phrase and
the portable mesh key remain local to the caller.
"""

from __future__ import annotations

import base64
import hmac
import inspect
import json
import os
import re
import unicodedata
from dataclasses import dataclass
from datetime import datetime, timezone
from hashlib import sha256
from typing import Any, Awaitable, Callable, Mapping, Protocol, TypeAlias, runtime_checkable

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

from .crypto import (
    _ECMASCRIPT_WHITESPACE_RE,
    CryptoProtocolError,
    canonical_portable_key,
    ecmascript_trim,
    json_stringify,
    text_encode,
)


BytesLike: TypeAlias = bytes | bytearray | memoryview
RandomBytes: TypeAlias = Callable[[int], bytes]

RECOVERY_ROOT_SALT = b"interocitor.recovery.root.v1"
LOCATOR_INFO = b"interocitor.recovery.locator.v1"
KEK_INFO = b"interocitor.recovery.kek.v1"
ROOT_ITERATIONS = 600_000
RECOVERY_FOLDER = "/.interocitor/recovery"
_LOCATOR_RE = re.compile(r"^[A-Za-z0-9_-]{43}$")


class RecoveryError(ValueError):
    """Raised when a phrase or recovery wrapper cannot be used safely."""


def _reject_non_json_constant(value: str) -> None:
    raise ValueError(f"Invalid JSON constant: {value}")


def _to_base64url(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).rstrip(b"=").decode("ascii")


def _from_base64url(value: str, *, field: str) -> bytes:
    if not isinstance(value, str):
        raise RecoveryError(f"Recovery wrapper {field} must be a string")
    padded = value + "=" * ((4 - len(value) % 4) % 4)
    try:
        # ``validate=True`` keeps malformed lookup capabilities from being
        # silently accepted by Python's otherwise permissive decoder.
        return base64.b64decode(padded.encode("ascii"), altchars=b"-_", validate=True)
    except (UnicodeEncodeError, ValueError) as error:
        raise RecoveryError(f"Recovery wrapper {field} is not valid base64url") from error


def normalize_recovery_phrase(phrase: str) -> str:
    """Apply core's NFKD, trim, and whitespace normalization."""

    if not isinstance(phrase, str):
        raise TypeError("phrase must be a string")
    return ecmascript_trim(_ECMASCRIPT_WHITESPACE_RE.sub(" ", unicodedata.normalize("NFKD", phrase)))


def _require_phrase(phrase: str) -> str:
    normalized = normalize_recovery_phrase(phrase)
    if not normalized:
        raise RecoveryError("Recovery phrase must not be empty")
    return normalized


def _derive_recovery_root(phrase: str) -> bytes:
    return PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=RECOVERY_ROOT_SALT,
        iterations=ROOT_ITERATIONS,
    ).derive(text_encode(_require_phrase(phrase)))


def _derive_locator(root: bytes) -> str:
    return _to_base64url(hmac.new(root, LOCATOR_INFO, sha256).digest())


def _derive_kek(root: bytes, salt: bytes) -> bytes:
    return HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        info=KEK_INFO,
    ).derive(root)


def _wrapper_aad(locator: str, salt: str) -> bytes:
    return f"interocitor.recovery.wrapper.v1|{locator}|{salt}".encode("utf-8")


def recovery_path(locator: str) -> str:
    """Return the generic adapter path used by core for a recovery wrapper."""

    if not _LOCATOR_RE.fullmatch(locator):
        raise RecoveryError("Invalid recovery locator")
    return f"{RECOVERY_FOLDER}/{locator}.json"


@dataclass(frozen=True)
class RecoveredMeshCredentials:
    """Portable mesh credentials recovered from a phrase-protected wrapper."""

    remote_path: str
    portable_key: str
    mesh_id: str | None = None

    def to_dict(self) -> dict[str, str]:
        value: dict[str, str] = {
            "remotePath": self.remote_path,
            "portableKey": self.portable_key,
        }
        if self.mesh_id:
            value["meshId"] = self.mesh_id
        return value

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "RecoveredMeshCredentials":
        if not isinstance(value, Mapping):
            raise RecoveryError("Recovery wrapper contains invalid credentials")
        remote_path = value.get("remotePath", value.get("remote_path"))
        portable_key = value.get("portableKey", value.get("portable_key"))
        mesh_id = value.get("meshId", value.get("mesh_id"))
        if not isinstance(remote_path, str) or not isinstance(portable_key, str):
            raise RecoveryError("Recovery wrapper contains invalid credentials")
        try:
            portable_key = canonical_portable_key(portable_key)
        except (CryptoProtocolError, TypeError) as error:
            raise RecoveryError("Recovery wrapper contains invalid credentials") from error
        return cls(
            remote_path=remote_path,
            portable_key=portable_key,
            mesh_id=mesh_id if isinstance(mesh_id, str) and mesh_id else None,
        )


@dataclass(frozen=True)
class RecoveryKdf:
    """Version-1 public KDF parameters from a recovery wrapper."""

    salt: str
    iterations: int = ROOT_ITERATIONS

    def to_dict(self) -> dict[str, object]:
        return {
            "root": {"name": "PBKDF2-HMAC-SHA-256", "iterations": self.iterations},
            "kek": {"name": "HKDF-SHA-256", "salt": self.salt},
        }

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "RecoveryKdf":
        if not isinstance(value, Mapping):
            raise RecoveryError("Invalid recovery wrapper")
        root = value.get("root")
        kek = value.get("kek")
        if not isinstance(root, Mapping) or not isinstance(kek, Mapping):
            raise RecoveryError("Invalid recovery wrapper")
        iterations = root.get("iterations")
        salt = kek.get("salt")
        if (
            root.get("name") != "PBKDF2-HMAC-SHA-256"
            or type(iterations) is not int
            or iterations != ROOT_ITERATIONS
            or kek.get("name") != "HKDF-SHA-256"
            or not isinstance(salt, str)
        ):
            raise RecoveryError("Invalid recovery wrapper")
        return cls(salt=salt, iterations=iterations)


@dataclass(frozen=True)
class RecoveryWrapper:
    """Version-1 recovery wrapper with camel-case wire serialization helpers."""

    v: int
    alg: str
    kdf: RecoveryKdf
    locator: str
    iv: str
    ciphertext: str
    created_at: str

    def to_dict(self) -> dict[str, object]:
        return {
            "v": self.v,
            "alg": self.alg,
            "kdf": self.kdf.to_dict(),
            "locator": self.locator,
            "iv": self.iv,
            "ciphertext": self.ciphertext,
            "createdAt": self.created_at,
        }

    def to_json(self) -> str:
        return json_stringify(self.to_dict())

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "RecoveryWrapper":
        if not isinstance(value, Mapping):
            raise RecoveryError("Invalid recovery wrapper")
        version = value.get("v")
        alg = value.get("alg")
        locator = value.get("locator")
        iv = value.get("iv")
        ciphertext = value.get("ciphertext")
        created_at = value.get("createdAt", value.get("created_at"))
        if (
            type(version) is not int
            or version != 1
            or alg != "AES-GCM"
            or not isinstance(locator, str)
            or not _LOCATOR_RE.fullmatch(locator)
            or not isinstance(iv, str)
            or not isinstance(ciphertext, str)
            or not isinstance(created_at, str)
        ):
            raise RecoveryError("Invalid recovery wrapper")
        return cls(
            v=version,
            alg=alg,
            kdf=RecoveryKdf.from_dict(value.get("kdf")),
            locator=locator,
            iv=iv,
            ciphertext=ciphertext,
            created_at=created_at,
        )

    @classmethod
    def from_json(cls, value: str | BytesLike) -> "RecoveryWrapper":
        if isinstance(value, str):
            text = value
        elif isinstance(value, (bytes, bytearray, memoryview)):
            text = bytes(value).decode("utf-8", errors="replace")
        else:
            raise TypeError("Recovery wrapper must be text or bytes-like")
        try:
            parsed = json.loads(text, parse_constant=_reject_non_json_constant)
        except (TypeError, ValueError, json.JSONDecodeError) as error:
            raise RecoveryError("Invalid recovery wrapper") from error
        if not isinstance(parsed, Mapping):
            raise RecoveryError("Invalid recovery wrapper")
        return cls.from_dict(parsed)


def _coerce_credentials(
    credentials: RecoveredMeshCredentials | Mapping[str, Any],
) -> RecoveredMeshCredentials:
    if isinstance(credentials, RecoveredMeshCredentials):
        return RecoveredMeshCredentials.from_dict(credentials.to_dict())
    return RecoveredMeshCredentials.from_dict(credentials)


def _coerce_wrapper(wrapper: RecoveryWrapper | Mapping[str, Any]) -> RecoveryWrapper:
    if isinstance(wrapper, RecoveryWrapper):
        # A caller can construct a dataclass directly, so run the wire-format
        # validator here just as core's unwrap function always calls assertWrapper.
        return RecoveryWrapper.from_dict(wrapper.to_dict())
    return RecoveryWrapper.from_dict(wrapper)


def _exact_random_bytes(random_bytes: RandomBytes, length: int) -> bytes:
    value = random_bytes(length)
    if not isinstance(value, (bytes, bytearray, memoryview)):
        raise TypeError("random_bytes must return bytes-like data")
    raw = bytes(value)
    if len(raw) != length:
        raise RecoveryError(f"random_bytes returned {len(raw)} bytes; expected {length}")
    return raw


def _created_at_now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def recovery_locator(phrase: str) -> str:
    """Derive the stable opaque locator for a normalized recovery phrase."""

    return _derive_locator(_derive_recovery_root(phrase))


def create_recovery_wrapper(
    phrase: str,
    credentials: RecoveredMeshCredentials | Mapping[str, Any],
    *,
    random_bytes: RandomBytes = os.urandom,
    created_at: str | None = None,
) -> RecoveryWrapper:
    """Encrypt portable mesh credentials under a client-provided phrase.

    ``random_bytes`` and ``created_at`` are injectable to support interoperability
    vectors; production callers should use their secure defaults.
    """

    recovered = _coerce_credentials(credentials)
    if not recovered.remote_path or not recovered.portable_key:
        raise RecoveryError("Recovery credentials require remotePath and portableKey")
    if created_at is not None and not isinstance(created_at, str):
        raise TypeError("created_at must be a string or None")

    root = _derive_recovery_root(phrase)
    locator = _derive_locator(root)
    salt = _exact_random_bytes(random_bytes, 16)
    iv = _exact_random_bytes(random_bytes, 12)
    salt_text = _to_base64url(salt)
    plaintext = json_stringify(recovered.to_dict()).encode("utf-8")
    ciphertext = AESGCM(_derive_kek(root, salt)).encrypt(
        iv,
        plaintext,
        _wrapper_aad(locator, salt_text),
    )
    return RecoveryWrapper(
        v=1,
        alg="AES-GCM",
        kdf=RecoveryKdf(salt=salt_text),
        locator=locator,
        iv=_to_base64url(iv),
        ciphertext=_to_base64url(ciphertext),
        created_at=created_at if created_at is not None else _created_at_now(),
    )


def unwrap_recovery_wrapper(
    phrase: str,
    wrapper: RecoveryWrapper | Mapping[str, Any],
) -> RecoveredMeshCredentials:
    """Validate and decrypt a core-compatible phrase recovery wrapper."""

    parsed = _coerce_wrapper(wrapper)
    root = _derive_recovery_root(phrase)
    locator = _derive_locator(root)
    if not hmac.compare_digest(locator, parsed.locator):
        raise RecoveryError("Recovery phrase does not match this wrapper")

    try:
        salt = _from_base64url(parsed.kdf.salt, field="kdf.kek.salt")
        iv = _from_base64url(parsed.iv, field="iv")
        ciphertext = _from_base64url(parsed.ciphertext, field="ciphertext")
        plaintext = AESGCM(_derive_kek(root, salt)).decrypt(
            iv,
            ciphertext,
            _wrapper_aad(parsed.locator, parsed.kdf.salt),
        )
        value = json.loads(
            plaintext.decode("utf-8", errors="replace"),
            parse_constant=_reject_non_json_constant,
        )
        return RecoveredMeshCredentials.from_dict(value)
    except RecoveryError as error:
        if str(error).startswith("Recovery wrapper contains"):
            raise
        # Match core's intentionally non-specific message for malformed
        # ciphertext, KDF material, or failed GCM authentication.
        raise RecoveryError("Recovery phrase could not unlock this wrapper") from error
    except Exception as error:
        raise RecoveryError("Recovery phrase could not unlock this wrapper") from error


@runtime_checkable
class RecoveryReader(Protocol):
    """The minimal adapter capability used by :func:`recover_mesh_credentials`."""

    def read_recovery_wrapper(self, locator: str) -> Awaitable[BytesLike]:
        """Read a serialized wrapper by its validated opaque locator."""


async def _maybe_await(value: Any) -> Any:
    if inspect.isawaitable(value):
        return await value
    return value


def _adapter_method(adapter: object, *names: str) -> Callable[..., Any] | None:
    for name in names:
        method = getattr(adapter, name, None)
        if callable(method):
            return method
    return None


async def recover_mesh_credentials(adapter: object, phrase: str) -> RecoveredMeshCredentials:
    """Read, validate, and unlock credentials through an async-capable adapter.

    Python adapters normally expose ``read_recovery_wrapper(locator)``.  The
    camel-case core spelling and the generic ``read_file`` fallback are accepted
    too, which makes this usable with a small storage adapter rather than a
    recovery-specific endpoint.
    """

    locator = recovery_locator(phrase)
    reader = _adapter_method(adapter, "read_recovery_wrapper", "readRecoveryWrapper")
    if reader is not None:
        data = await _maybe_await(reader(locator))
    else:
        read_file = _adapter_method(adapter, "read_file", "readFile")
        if read_file is None:
            raise TypeError("adapter must provide read_recovery_wrapper or read_file")
        data = await _maybe_await(read_file(recovery_path(locator)))

    try:
        wrapper = RecoveryWrapper.from_json(data)
    except (TypeError, RecoveryError) as error:
        raise RecoveryError("Invalid recovery wrapper") from error
    return unwrap_recovery_wrapper(phrase, wrapper)


async def publish_recovery_wrapper(adapter: object, wrapper: RecoveryWrapper | Mapping[str, Any]) -> None:
    """Serialize and write a wrapper through a recovery or generic adapter."""

    parsed = _coerce_wrapper(wrapper)
    data = parsed.to_json().encode("utf-8")
    writer = _adapter_method(adapter, "write_recovery_wrapper", "writeRecoveryWrapper")
    if writer is not None:
        await _maybe_await(writer(parsed.locator, data))
        return

    ensure_folder = _adapter_method(adapter, "ensure_folder", "ensureFolder")
    write_file = _adapter_method(adapter, "write_file", "writeFile")
    if ensure_folder is None or write_file is None:
        raise TypeError("adapter must provide write_recovery_wrapper or ensure_folder and write_file")
    await _maybe_await(ensure_folder(RECOVERY_FOLDER))
    await _maybe_await(write_file(recovery_path(parsed.locator), data))

File: packages/core/src/crypto/recovery.ts
// compass: interocitor.trust.recovery

/**
 * Recovery-key protocol for portable-key meshes.
 *
 * An application-provided recovery phrase never leaves the client. It is
 * converted into a recovery root, which deterministically produces an opaque
 * lookup locator and derives a KEK for encrypted mesh credentials.
 */

import type { StorageAdapter } from "../core/types.ts";

const encoder = new TextEncoder();
const decoder = new TextDecoder();
const RECOVERY_ROOT_SALT = encoder.encode("interocitor.recovery.root.v1");
const LOCATOR_INFO = encoder.encode("interocitor.recovery.locator.v1");
const KEK_INFO = encoder.encode("interocitor.recovery.kek.v1");
const ROOT_ITERATIONS = 600_000;
const RECOVERY_FOLDER = "/.interocitor/recovery";
const LOCATOR_RE = /^[A-Za-z0-9_-]{43}$/;

export interface RecoveredMeshCredentials {
  /** Remote root passed back to `Interocitor` as `remotePath`. */
  remotePath: string;
  /** High-entropy base58 material consumed by `PortablePassphraseKeySource`. */
  portableKey: string;
  /** Manifest mesh identity when it was known while creating the wrapper. */
  meshId?: string;
}

/** Serialized encrypted recovery record stored by a remote adapter. */
export interface RecoveryWrapper {
  /** Recovery-wrapper format version. */
  v: 1;
  /** Authenticated-encryption algorithm used for `ciphertext`. */
  alg: "AES-GCM";
  /** Public derivation parameters required to reproduce the wrapping key. */
  kdf: {
    root: {
      name: "PBKDF2-HMAC-SHA-256";
      /** Fixed work factor accepted by this format version. */
      iterations: number;
    };
    kek: {
      name: "HKDF-SHA-256";
      /** Per-wrapper random salt encoded as base64url. */
      salt: string;
    };
  };
  /** Phrase-derived, base64url lookup capability; it does not contain the mesh ID. */
  locator: string;
  /** Random 96-bit AES-GCM IV encoded as base64url. */
  iv: string;
  /** Authenticated encrypted `RecoveredMeshCredentials` JSON, encoded as base64url. */
  ciphertext: string;
  /** ISO timestamp recorded by the creating client. */
  createdAt: string;
}

/**
 * Adapter extension for services that keep recovery wrappers outside a known
 * mesh path. Generic adapters use `/.interocitor/recovery/` instead.
 */
export interface RecoveryStorageAdapter {
  /** Return the serialized wrapper for an already validated locator. */
  readRecoveryWrapper(locator: string): Promise<Uint8Array>;
  /** Store one serialized wrapper under an already validated locator. */
  writeRecoveryWrapper(locator: string, data: Uint8Array): Promise<void>;
}

function toBase64Url(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) binary += String.fromCodePoint(byte);
  return btoa(binary).replaceAll("+", "-").replaceAll("/", "_").replaceAll("=", "");
}

function fromBase64Url(value: string): Uint8Array {
  const padded =
    value.replaceAll("-", "+").replaceAll("_", "/") + "=".repeat((4 - (value.length % 4)) % 4);
  const binary = atob(padded);
  return Uint8Array.from(binary, (char) => char.codePointAt(0)!);
}

function asBuffer(bytes: Uint8Array): ArrayBuffer {
  return bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength) as ArrayBuffer;
}

function normalizePhrase(phrase: string): string {
  return phrase.normalize("NFKD").trim().replaceAll(/\s+/g, " ");
}

function assertPhrase(phrase: string): string {
  const normalized = normalizePhrase(phrase);
  if (!normalized) throw new Error("Recovery phrase must not be empty");
  return normalized;
}

function assertWrapper(value: unknown): asserts value is RecoveryWrapper {
  if (!value || typeof value !== "object") throw new Error("Invalid recovery wrapper");
  const wrapper = value as Partial<RecoveryWrapper>;
  if (
    wrapper.v !== 1 ||
    wrapper.alg !== "AES-GCM" ||
    wrapper.kdf?.root?.name !== "PBKDF2-HMAC-SHA-256" ||
    wrapper.kdf.root.iterations !== ROOT_ITERATIONS ||
    wrapper.kdf.kek?.name !== "HKDF-SHA-256" ||
    typeof wrapper.kdf.kek.salt !== "string" ||
    !LOCATOR_RE.test(String(wrapper.locator || "")) ||
    typeof wrapper.iv !== "string" ||
    typeof wrapper.ciphertext !== "string" ||
    typeof wrapper.createdAt !== "string"
  ) {
    throw new Error("Invalid recovery wrapper");
  }
}

async function deriveRecoveryRoot(phrase: string): Promise<Uint8Array> {
  const phraseBytes = encoder.encode(assertPhrase(phrase));
  const source = await crypto.subtle.importKey(
    "raw",
    phraseBytes.buffer as ArrayBuffer,
    "PBKDF2",
    false,
    ["deriveBits"],
  );
  const bits = await crypto.subtle.deriveBits(
    { name: "PBKDF2", hash: "SHA-256", salt: RECOVERY_ROOT_SALT, iterations: ROOT_ITERATIONS },
    source,
    256,
  );
  return new Uint8Array(bits);
}

async function deriveLocator(root: Uint8Array): Promise<string> {
  const key = await crypto.subtle.importKey(
    "raw",
    root.buffer as ArrayBuffer,
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  return toBase64Url(new Uint8Array(await crypto.subtle.sign("HMAC", key, LOCATOR_INFO)));
}

async function deriveKek(root: Uint8Array, salt: Uint8Array): Promise<CryptoKey> {
  const source = await crypto.subtle.importKey("raw", root.buffer as ArrayBuffer, "HKDF", false, [
    "deriveKey",
  ]);
  return crypto.subtle.deriveKey(
    { name: "HKDF", hash: "SHA-256", salt: asBuffer(salt), info: asBuffer(KEK_INFO) },
    source,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"],
  );
}

function wrapperAad(locator: string, salt: string): Uint8Array {
  return encoder.encode(`interocitor.recovery.wrapper.v1|${locator}|${salt}`);
}

function recoveryPath(locator: string): string {
  if (!LOCATOR_RE.test(locator)) throw new Error("Invalid recovery locator");
  return `${RECOVERY_FOLDER}/${locator}.json`;
}

function recoveryAdapter(adapter: StorageAdapter): RecoveryStorageAdapter | null {
  const candidate = adapter as StorageAdapter & Partial<RecoveryStorageAdapter>;
  return typeof candidate.readRecoveryWrapper === "function" &&
    typeof candidate.writeRecoveryWrapper === "function"
    ? (candidate as RecoveryStorageAdapter)
    : null;
}

/**
 * Derive the opaque remote lookup identifier for a recovery phrase.
 *
 * Whitespace is normalized with NFKD before derivation. Rejects an empty
 * phrase. The returned locator is stable for the same normalized phrase and
 * can be used to test phrase guesses against a copied wrapper.
 */
export async function recoveryLocator(phrase: string): Promise<string> {
  return deriveLocator(await deriveRecoveryRoot(phrase));
}

/**
 * Encrypt portable mesh credentials under an application-provided phrase.
 *
 * This creates the wrapper in memory; call `publishRecoveryWrapper` to
 * store it. Rejects an empty phrase or missing `remotePath`/`portableKey`.
 * Phrase generation and validation remain application responsibilities.
 */
export async function createRecoveryWrapper(
  phrase: string,
  credentials: RecoveredMeshCredentials,
): Promise<RecoveryWrapper> {
  if (!credentials.remotePath || !credentials.portableKey) {
    throw new Error("Recovery credentials require remotePath and portableKey");
  }
  const root = await deriveRecoveryRoot(phrase);
  const [locator, salt] = await Promise.all([
    deriveLocator(root),
    Promise.resolve(crypto.getRandomValues(new Uint8Array(16))),
  ]);
  const saltText = toBase64Url(salt);
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const kek = await deriveKek(root, salt);
  const plaintext: RecoveredMeshCredentials = {
    remotePath: credentials.remotePath,
    portableKey: credentials.portableKey,
    ...(credentials.meshId ? { meshId: credentials.meshId } : {}),
  };
  const ciphertext = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: asBuffer(iv), additionalData: asBuffer(wrapperAad(locator, saltText)) },
    kek,
    encoder.encode(JSON.stringify(plaintext)),
  );
  return {
    v: 1,
    alg: "AES-GCM",
    kdf: {
      root: { name: "PBKDF2-HMAC-SHA-256", iterations: ROOT_ITERATIONS },
      kek: { name: "HKDF-SHA-256", salt: saltText },
    },
    locator,
    iv: toBase64Url(iv),
    ciphertext: toBase64Url(new Uint8Array(ciphertext)),
    createdAt: new Date().toISOString(),
  };
}

/**
 * Validate and decrypt a recovery wrapper with the supplied phrase.
 *
 * Rejects unsupported/malformed wrappers, a phrase whose locator differs, an
 * authentication failure, or decrypted data without the required credentials.
 */
export async function unwrapRecoveryWrapper(
  phrase: string,
  wrapper: RecoveryWrapper,
): Promise<RecoveredMeshCredentials> {
  assertWrapper(wrapper);
  const root = await deriveRecoveryRoot(phrase);
  const locator = await deriveLocator(root);
  if (locator !== wrapper.locator) throw new Error("Recovery phrase does not match this wrapper");
  try {
    const plaintext = await crypto.subtle.decrypt(
      {
        name: "AES-GCM",
        iv: asBuffer(fromBase64Url(wrapper.iv)),
        additionalData: asBuffer(wrapperAad(wrapper.locator, wrapper.kdf.kek.salt)),
      },
      await deriveKek(root, fromBase64Url(wrapper.kdf.kek.salt)),
      asBuffer(fromBase64Url(wrapper.ciphertext)),
    );
    const value = JSON.parse(decoder.decode(plaintext)) as Partial<RecoveredMeshCredentials>;
    if (!value || typeof value.remotePath !== "string" || typeof value.portableKey !== "string") {
      throw new Error("Recovery wrapper contains invalid credentials");
    }
    return {
      remotePath: value.remotePath,
      portableKey: value.portableKey,
      ...(typeof value.meshId === "string" && value.meshId ? { meshId: value.meshId } : {}),
    };
  } catch (error) {
    if (error instanceof Error && error.message.startsWith("Recovery wrapper contains"))
      throw error;
    throw new Error("Recovery phrase could not unlock this wrapper", { cause: error });
  }
}

/**
 * Serialize and upload a validated wrapper.
 *
 * Adapters implementing `RecoveryStorageAdapter` choose the recovery
 * endpoint. Generic adapters, including WebDAV, write beneath
 * `/.interocitor/recovery/`. Overwrite behavior belongs to that adapter.
 */
export async function publishRecoveryWrapper(
  adapter: StorageAdapter,
  wrapper: RecoveryWrapper,
): Promise<void> {
  assertWrapper(wrapper);
  const data = encoder.encode(JSON.stringify(wrapper));
  const recovery = recoveryAdapter(adapter);
  if (recovery) return recovery.writeRecoveryWrapper(wrapper.locator, data);
  await adapter.ensureFolder(RECOVERY_FOLDER);
  await adapter.writeFile(recoveryPath(wrapper.locator), data);
}

/**
 * Locate, download, validate, and decrypt portable mesh credentials.
 *
 * The caller still supplies an adapter that can authenticate to the storage
 * provider. Rejects remote read failures, malformed wrappers, and phrases that
 * cannot authenticate the stored ciphertext.
 */
export async function recoverMeshCredentials(
  adapter: StorageAdapter,
  phrase: string,
): Promise<RecoveredMeshCredentials> {
  const locator = await recoveryLocator(phrase);
  const recovery = recoveryAdapter(adapter);
  const data = recovery
    ? await recovery.readRecoveryWrapper(locator)
    : await adapter.readFile(recoveryPath(locator));
  let wrapper: unknown;
  try {
    wrapper = JSON.parse(decoder.decode(data));
  } catch {
    throw new Error("Invalid recovery wrapper");
  }
  assertWrapper(wrapper);
  return unwrapRecoveryWrapper(phrase, wrapper);
}
